---
title: "Material Semana Noviembre 19 2021"
output: github_document
---
```{r}
# Set working directory
#getwd()
setwd("/Users/kalilamali/Documents/cursos/ucm_genomes_and_genome_analysis/S4_Transcriptomics/")
```

*1. Load gene quantifications from alignment:*
```{r}
# Install required packages
# install.packages("matrixStats")
# BiocManager::install("DESeq2")
# library(DESeq2)
# install.packages("pheatmap")
# 
# install.packages("igraph", type = "binary")
# if(!requireNamespace("remotes", quietly = TRUE))
#   install.packages("remotes")
# remotes::install_github("YuLab-SMU/clusterProfiler")
# 
# install.packages("tidyverse")
# BiocManager::install("org.Hs.eg.db")
```
Gene quantifications will result after completing an alignment, quantifications can be calculated at the gene, transcript or exon level.
```{r}
# Download raw counts per gene
# wget https://www.dropbox.com/s/91u3vk7fvrlk2jy/GSE50957_rnaseq_processed_data.RData

# Load data
load("GSE50957_rnaseq_processed_data.RData")

# Observe data
head(raw.gene.counts)
head(sample.table)

# Review if content in data is identical
identical(colnames(raw.gene.counts),rownames(sample.table))

# Select the gene counts for which we have metadata
raw.gene.counts <- raw.gene.counts[,rownames(sample.table)]

# Rename data
sratable.sel <- sample.table
cts.sel <- raw.gene.counts

# Change data format
cts.sel <- as.matrix(cts.sel)

# the count data generated by RNA-seq exhibits overdispersion (variance > mean) and the statistical distribution used to model the counts needs to account for this overdispersion.
# Load package
library(matrixStats)
# This plot refers to the gene counts
plot(x = log(rowMeans(cts.sel)), y = log(rowVars(cts.sel)),
     pch=".", col="grey")
lines(x=0:100, y=0:100, col="red")
```
*2. Differential expression analysis using DESeq2*
We will complete a differential expression analysis using a paired analysis. In this analysis we will control for variation associated with individual variation using the factor “subj_id”. The main condition for our contrast would be the infection status which is determined by “time_point”.
```{r}
# Load package
library(DESeq2)

# Controling for variation in the subjects "subj_id"
# We are contrasting infection status which is determined by “time_point”
dds <- DESeqDataSetFromMatrix(countData = cts.sel,
                              colData = sratable.sel,
                              design = ~ subj_id + time_point)
# Remove empty rows
dds <- dds[ rowSums(counts(dds)) > 1, ]

dds <- DESeq(dds)

plotDispEsts(dds)

```
Now we can check the result of the contrast defining our threshold of significance using the adjusted P-value, also know as False Discovery Rate (FDR). In this case we will use a 5% of FDR or 0.05 asjusted pvalue as the limit to reject the null hypothesis.
```{r}
# Check result names
resultsNames(dds)
# Threshold with pvalue
res <- results(dds, contrast = c("time_point", "Pre", "Post"), alpha = 0.05)

# LFC stands for logFoldChange to see which ones are expressed and which ones are not.
summary(res)

# Save the results in a RData file
save(dds, res, cts.sel, sratable.sel, file = "results_DEG_GSE50957_all_genes_more_than_1.RData")
```
Plot Counts of any gene using PlotCounts function from DESeq2. Plot the read counts per gene, normalized comparing across conditions:
```{r}
plotCounts(dds, gene="ENSG00000000457.13", intgroup = "time_point")
```
*3. Exploratory data analysis (EDA)*
3.1 Quality Control (QC) of libraries.
To estimate library sizes, we have to sum all aligned reads per sample.
```{r}
colSums(counts(dds))
```
Library sizes plots:
```{r}
# Sum gene counts for each sample and order it
sorted.cts.sel <- sort(colSums(cts.sel))/1E6
# Order the other table acording the sample order above
sratable.sel.2 <- sratable.sel[names(sorted.cts.sel),]

# Plot library sizes per sample
barplot(sorted.cts.sel,
        horiz = F,
        cex.main= 1,
        main="Library sizes", # Title
        col=ifelse(sratable.sel.2$time_point=="Pre", "blue", "red"))
```
3.2 Gene expresion distribution and MA plots.
For mean expression distribution across all samples:
```{r}
plot(density(log2(res$baseMean)))
```
For MA plot:
```{r}
plotMA(res)
```
3.3 Principal Component Analysis (PCA).
We can devise the main source of variation in our data using PCA of the top variable genes in our dataset. The selection of top variable genes MUST be independent of the differential expression results.
```{r}
## Variant Stabilization Dispersion and Principal Component Analysis

# Do the vts: variant stabilization dispersion
vsd <- vst(dds)
vst.counts <- assay(vsd)
# Do the PCA
DESeq2::plotPCA(vsd, intgroup = c("time_point"), ntop=1000)
```
3.4 Heatmap of top variable genes
```{r}
## Heatmap of highly variable genes
library(matrixStats)

# Sort the vst
top.genes <- vst.counts[order(rowVars(vst.counts), decreasing = T)[1:1000],]

# Load the package
library(pheatmap)

# Plot
pheatmap(top.genes,
         annotation = as.data.frame(colData(dds)),# filename = "Heatmap.pdf",
         scale="row", show_rownames = FALSE)
```
*4. Functional enrichment analysis using Gene Ontology.*
Gene Ontoloty: http://geneontology.org
```{r}
## Enrichment analysis of over-represented GO terms
# Load packages
library(clusterProfiler)
library(ggplot2)
library(org.Hs.eg.db)

ego <- enrichGO(gene          = substr(rownames(subset(res, padj < 0.05)),1,15),
                universe      = substr(rownames(res),1,15),
                keyType       = 'ENSEMBL',
                OrgDb         = org.Hs.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = TRUE)


pdf(file = "dotplot_Pre_vs_Post_malaria.pdf", width = 30, height = 30)
dotplot(ego, showCategory=200) + ggtitle("dotplot for Pre_vs_Post")
dev.off()
```
# Tutorial:
https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
# Significado de plots:
https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html

# Homeworks:
*Exercise 1: Which is the gene with higher absolute fold change in your chromosome? Hint: You should use BioMart to obtain the corresponding chromosome of genes and, it would be preferable to use ENSEMBL Gene ID stable.*

Results tables are generated using the function results, which extracts a results table with log2 fold changes, p values and adjusted p values. 
```{r}
# Select a BioMart database and dataset
library("biomaRt")
my.ensembl.mart = useEnsembl("ENSEMBL_MART_ENSEMBL")
hs.dataset = useDataset("hsapiens_gene_ensembl", mart=my.ensembl.mart)

# Build a BioMart query
available.filters = listFilters(hs.dataset)
head(available.filters)
available.attributes = listAttributes(hs.dataset)
available.attributes <- subset(available.attributes, page!="homologs")
head(cbind(available.attributes[1:(360/2),], available.attributes[((360/2)+1):360,]))
# Query
myvals = c('21')
mycols = c('ensembl_gene_id_version','chromosome_name')
anndf <- getBM(attributes=mycols,
      filters = 'chromosome_name', 
      values = myvals, 
      mart = hs.dataset)

# Get the results from the differential expression analysis (DEA)
#dds <- DESeq(dds)
#res <- results(dds)
res

# Get the absolute fold change
#The opposite of log2(x) is 2^(x)
resdf <- data.frame(res)
resdf$FoldChange <- abs(2^(resdf$log2FoldChange))
resdf$ensembl_gene_id_version <- rownames(resdf)
#resdf$ensembl_gene_id_version <- gsub("\\.","",resdf$ensembl_gene_id_version)

# Join both dfs
newdf <- merge(resdf, anndf, by='ensembl_gene_id_version')

# Sort column FoldChange from more to less
highfcdf<- newdf[order(-newdf$FoldChange),]
highfcdf
```
# ENSG00000227999.1

*Exercise 2.1: Which is the gene that is strongly differentially expressed (lowest adjusted P-value) in your chromosome? Provide the ENSEMBL Gene ID stable.*
```{r}
# Sort column padj from less to more
lowpadjdf <- newdf[order(newdf$padj),]
lowpadjdf
```
# ENSG00000228318.3

*Exercise 2.2: What is its mean expression across all samples? Use normalized counts to estimate the mean expression using the function “counts” (e.g. counts(dds, normalized=TRUE) ) from DESeq2 to retrieve a matrix with normalized counts of the gene expression matrix.*
```{r}
# Get the normalized counts
norm.gene.counts <- counts(dds, normalized=TRUE)
# Change the format to a matrix to do stats
ncts.sel <- as.matrix(norm.gene.counts)
# Get the mean across the means of columns because each column is a sample
mean(colMeans(ncts.sel))
```
*Exercise 3: What is the mean expression across pre-infection samples of the gene that is strongly differentially expressed (lowest adjusted P-value) in your chromosome? Use normalized gene counts in a similar way as the previous exercise.*
```{r}
# Gene from Exercise 2.1.
gene <- 'ENSG00000228318.3'

#Get rows where time_point column is == to "Pre"
sample.table.pre <- subset(sample.table, time_point=="Pre")

# Select the gene counts for the pre samples only
ncts.sel <- ncts.sel[,rownames(sample.table.pre)]

# Get the mean
mean(ncts.sel[gene,])
```
*Exercise 4: How many differentially expressed genes for this threshold FDR < 20% (or adjusted P-value < 0.2) are in your chromosome?*
```{r}
# Threshold with pvalue
res20 <- results(dds, contrast = c("time_point", "Pre", "Post"), alpha = 0.2)
# View results summary
summary(res20)
# Create df of results
res20df <- data.frame(res20)
# Put gene names as a col before joining
res20df$ensembl_gene_id_version <- rownames(res20df)
# Join both dfs
newdf <- merge(res20df, anndf, by='ensembl_gene_id_version')
dim(newdf)
```
# 432

*Exercise 5: Reproduce a Volcano Plot (x.axis=log2 Fold Change and y-axis= -log10(P-value) and color as “red” differentially expressed genes (FDR < 10%) located within your chromosome and the rest of genes as grey. Save the Volcano Plot in PDF for (5 x 5 inches).*
A volcano plot is a type of scatter plot represents differential expression of features (genes for example): on the x-axis we typically find the fold change and on the y-axis the p-value.
#Plot tutorial:
https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html
```{r}
# Load package
library(ggplot2)

# Process data
# remove rows that contain NA values
de <- newdf[complete.cases(newdf), ]
# Create a column to get differentially express genes (p<0.1)
de$diffexpressed <- "NO"
# if pvalue < 0.1, set as "YES" 
de$diffexpressed[de$pvalue < 0.1] <- "YES"

# Plot
p <- ggplot(data=de, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed)) + geom_point() + theme_minimal()
# Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold 
#p2 <- p + geom_vline(xintercept=c(-0.6, 0.6), col="red") +
#        geom_hline(yintercept=-log10(0.1), col="red")
# Change point color 
# By default, it is assigned to the categories in an alphabetical order):
p3 <- p + scale_color_manual(values=c("grey", "red"))
p3

# Save plot
# In the console type one by one in order
#pdf(file="volcano_plot_s4.pdf", width=5, height=5)
#p3
#dev.off()
```

*Exercise 6.1: How many genes annotated to your for your assigned GO term are differentially expressed (FDR < 10%)? Check which GO term has been assigned to you in this Google Sheet link*
```{r}
# Query
anndf = getBM(attributes=c('ensembl_gene_id_version'),
      filters="go",
      values="GO:0005856", 
      mart=hs.dataset)

# Threshold with pvalue
res10 <- results(dds, contrast = c("time_point", "Pre", "Post"), alpha = 0.1)
# View results summary
summary(res10)
# Create df of results
res10df <- data.frame(res10)
# Put gene names as a col before joining
res10df$ensembl_gene_id_version <- rownames(res10df)
# Join both dfs
newdf <- merge(res10df, anndf, by='ensembl_gene_id_version')
dim(newdf)
```
*Exercise 6.2: What is the name of your assigned GO term?*
Go to the webpage: http://geneontology.org/
cytoskeleton (GO:0005856)

*Exercise 7: What is the ratio between differentially expressed genes (FDR < 5%) annotated to the GO term that has been assigned to you and genes from the appropiate background or universe for functional analysis? Hint: Use as a reference the same background used in the tutorial.*
```{r}
# Query
anndf = getBM(attributes=c('ensembl_gene_id_version'),
      filters="go",
      values="GO:0005856", 
      mart=hs.dataset)

# Threshold with pvalue
res5 <- results(dds, contrast = c("time_point", "Pre", "Post"), alpha = 0.05)
# Make df
res5df <- data.frame(res5)
# Put gene names as a col before joining
res5df$ensembl_gene_id_version <- rownames(res5df)
# Join both dfs
newdf <- merge(res5df, anndf, by='ensembl_gene_id_version')
row.names(newdf) <- newdf$ensembl_gene_id_version
newdf[1] <- NULL
dim(newdf)

# Enrichment
ego <- enrichGO(gene          = substr(rownames(subset(newdf, padj < 0.05)),1,15),
                universe      = substr(rownames(res5),1,15),
                keyType       = 'ENSEMBL',
                OrgDb         = org.Hs.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = TRUE)


pdf(file = "dotplot_Pre_vs_Post_malaria.pdf", width = 30, height = 30)
dotplot(ego, showCategory=200) + ggtitle("dotplot for Pre_vs_Post")
dev.off()

# Get the number of genes from newdf
# Get the number of genes from universe (the background genes)
ratio = 85/length(ego@universe)
ratio
```

*Exercise 8.1: Could be possible to study Transcript Differential Expression from RNAseq data of samples from a library of a specific cell line cultured at different experimental conditions?*
```{r}

```

*Exercise 8.2: Justify your response.*
